Pass inductor strides forward in ddp optimizer

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120727
* __->__ #120523

            # Note: Returning Fake Tensors on First AOT Autograd Call
            #
            # Inductor will optimize strides of outputs when it deems it profitable.
            # For instance, converting to channels last. When we split the graph here
            # into multiple inductor compilations, we need to make sure that the
            # output strides of one compilation is appropriately passed to the subsequent
            # compilations. However, the mapping from inductor output to dynamo output
            # is non-trivial due to aot_autograd's deduping, de-aliasing, mutation, re-writing,
            # subclass handling, etc. In order to replay all this logic we set a flag such that
            # the first invocation of inductor in aot_autograd will return Fake Tensors with
            # appropriate strides. Then, all of aot autograd's runtime logic is replayed.
            # This gives us the appropriately strided outputs here which will reflect runtime strides.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

eellison committed:
Pass inductor strides forward in ddp optimizer

[ghstack-poisoned]

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison commented on torch/_dynamo/backends/distributed.py:
```diff
@@ -322,7 +329,18 @@ def run_node(self, n: Node) -> Any:
             # and these need to be FakeTensors, so we execute the module under fake_mode
             with self.fake_mode:
-                return curr_submod(*new_args, **kwargs)
+                # TODO - make sure the first outputs of inductor match with user visible outputs
```
@bdhirsh is this correct ?

eellison replied:
What I need is a mapping from dynamo output index -> aot autograd output index..

bdhirsh replied:
Does aot autograd output here == the outputs of the forward FX graph that we send to inductor?

The calling convention is kind of complicated :/ ideally we would factor it into a shared util that we can re-use here, but it should look something like:

(*mutated_inputs_that_go_outside_the_graph, *fw_outs, *fw_intermediate_bases, *saved_tensors, *saved_symints)

Two more complications are:

(1) subclasses make this worse, because dynamo sees subclass outputs, but the aot graph sees plain tensor outputs. So if you have subclasses that desugar into multiple tensors, the calling convention gets more complicated (you need to __tensor_unflatten__ the "buckets" of inner tensors back into each subclass output). Code for that here:  https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py#L781

(2) @angelayi is working on adding effect tokens to functionalization, to properly functionalize torchbind object op calls. That will be another calling convention change (the effect tokens are extra tensor inputs/outputs in the graph that dynamo does not see.

bdhirsh replied:
tbh - given all of the above, and given that ddp_optimizer is a "temporary" solution, I wonder if it would be safer to ban some of the above cases in ddp_optimizer rather than try to support all possible calling convention changes that AOTAutograd makes (and risk them diverging later).

Examples of things we could ban:

(1) subclasses (? probably ok, unclear if someone will come to us asking for ddp + subclasses to work before compiled ddp is more hardened)

(2) intermediates bases (pretty rare)

(3) input mutations (also... unclear? input mutations that we keep outside the graph are ideally relatively rare)

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


Pass along the output strides from inductor compilation instead of those computed from FakeTensor. I think we can delete lazy_ddp with these changes.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison commented on torch/_dynamo/backends/distributed.py:
```diff
@@ -322,7 +329,18 @@ def run_node(self, n: Node) -> Any:
-                return curr_submod(*new_args, **kwargs)
+                # TODO - make sure the first outputs of inductor match with user visible outputs
+                # (I think thats the case)
+                out = curr_submod(*new_args, **kwargs)
```
May need to handle non tensor outputs here, will revise

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


Pass along the output strides from inductor compilation instead of those computed from FakeTensor. I think we can delete lazy_ddp with these changes.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


Pass along the output strides from inductor compilation instead of those computed from FakeTensor. I think we can delete lazy_ddp with these changes.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison commented on torch/_guards.py:
```diff
@@ -708,15 +710,24 @@ def current_frame(frame_summary):
     @contextlib.contextmanager
     def report_output_strides():
```
Dont actually need these changes anymore.. Maybe cleaner this way? either way, can delete

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


Pass along the output strides from inductor compilation instead of those computed from FakeTensor. I think we can delete lazy_ddp with these changes.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison commented on torch/_guards.py:
```diff
@@ -618,6 +618,10 @@ def __init__(self, fake_mode):
 
+        # If this true, Aot Autograd will return output Fake Tensors with appropiate
+        # meta on the first invocation
+        self.fakify_first_call = False
```
Happy to put this somewhere else/rename.. Like output_strides, kind of a pain to thread through everything, and hopefully DDP Optimizer is somewhat temporary??

wconstab replied:
haha that's what i was saying last year.  but hopefully for real this time :)

bdhirsh replied:
Can't think of a better place for it either :)

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


Pass along the output strides from inductor compilation instead of those computed from FakeTensor. I think we can delete lazy_ddp with these changes.

Edit: it's non-trivial to propagate inductor strides to dynamo output strides because of things like Subclasses, aliasing deduplication, etc etc.

I have updated this pr so that we invoked aot_autograd once after compilation with fake tensors in order to propagate output strides. The actually inductor generated fn is skipped on this first call.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


Pass along the output strides from inductor compilation instead of those computed from FakeTensor. I think we can delete lazy_ddp with these changes.

Edit: it's non-trivial to propagate inductor strides to dynamo output strides because of things like Subclasses, aliasing deduplication, etc etc.

I have updated this pr so that we invoked aot_autograd once after compilation with fake tensors in order to propagate output strides. The actually inductor generated fn is skipped on this first call.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:
```diff
@@ -309,6 +315,23 @@ def aot_dispatch_autograd(
             if not hasattr(compiled_fw_func, "_boxed_call"):
                 compiled_fw_func = make_boxed_func(compiled_fw_func)
 
+            if tracing_context and tracing_context.fakify_first_call:
```
A giant `Note [Calling AOTAutograd runtime code once with FakeTensors]` or something describing the situation and why we need it here would be really nice.

eellison replied:
will add when we are aligned

yf225 replied:
+1 would be really nice to have this note

eellison replied:
Added

bdhirsh commented on torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:
```diff
@@ -309,6 +315,23 @@ def aot_dispatch_autograd(
+                        if not isinstance(out[i], Tensor):
+                            continue
+                        if all(
+                            guard_size_oblivious(s1 == s2)
```
🤔 I guess we're effectively specializing (and adding guards) on whether or not inductor generated different output strides. Are you worried this will add a bunch of extra guards though, if we have a graph with 100+ outputs?

eellison replied:
I dont want to add any extra guards here.. Let me switch to `suppress_guards`

bdhirsh commented on torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:
```diff
@@ -309,6 +315,23 @@ def aot_dispatch_autograd(
+                            continue
+                        out[i] = out[i].as_strided(out[i].shape, fwd_output_strides[i])
+
+                fakified_out = out
```
Hmm I'm actually a bit confused - if we're grabbing the FakeTensor graph outputs (with the proper inductor strides) from here, why do we need to bother invoking the AOTAutograd runtime code at all in order to send this info to ddp_optimizer? Why not stash this immediately in the tracing context and grab it?

I'm mostly curious because I'm wondering if I misunderstood something. It seems nice to avoid the "return fake tensors at runtime" situation if we can avoid it, since it does feel a bit like a footgun (we'd better make sure we never actually hit that codepath from when the user calls the runtime code).

eellison replied:
This is what I did in the first PR. My understanding is these outputs from inductor have no easy mapping to what is returned in dynamo.

yf225 replied:
curious: with changes in this PR, is the output of first iteration going to be fake tensors when DDPOptimizer is enabled?

eellison replied:
The first invocation will be, which occurs in the graph splitter

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


Pass along the output strides from inductor compilation instead of those computed from FakeTensor. I think we can delete lazy_ddp with these changes.

Edit: it's non-trivial to propagate inductor strides to dynamo output strides because of things like Subclasses, aliasing deduplication, etc etc.

I have updated this pr so that we invoked aot_autograd once after compilation with fake tensors in order to propagate output strides. The actually inductor generated fn is skipped on this first call.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


Pass along the output strides from inductor compilation instead of those computed from FakeTensor. I think we can delete lazy_ddp with these changes.

Edit: it's non-trivial to propagate inductor strides to dynamo output strides because of things like Subclasses, aliasing deduplication, etc etc.

I have updated this pr so that we invoked aot_autograd once after compilation with fake tensors in order to propagate output strides. The actually inductor generated fn is skipped on this first call.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


Pass along the output strides from inductor compilation instead of those computed from FakeTensor. I think we can delete lazy_ddp with these changes.

Edit: it's non-trivial to propagate inductor strides to dynamo output strides because of things like Subclasses, aliasing deduplication, etc etc.

I have updated this pr so that we invoked aot_autograd once after compilation with fake tensors in order to propagate output strides. The actually inductor generated fn is skipped on this first call.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

yf225 commented on torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:
```diff
@@ -428,20 +454,27 @@ def forward(ctx, *deduped_flat_tensor_args):
+            else:
+                nonlocal fakified_out
+                CompiledFunction._fakify_first_call = False
+                fw_outs = fakified_out
```
to be safe, should we assert `fakified_out is not None` before this line?

kadeng commented:
There's a torchbench test failure with the moco model. I copied the relevant error message: P1189842447 

Among the error messages is something that looks like some function/method does not like fake tensors: **RuntimeError: The tensor has a non-zero number of elements, but its data is not allocated yet**. Caffe2 uses a lazy allocation, so you will need to call mutable_data() or raw_mutable_data() to actually allocate memory.

kadeng commented on torch/_dynamo/backends/distributed.py:
```diff
@@ -321,14 +336,20 @@ def run_node(self, n: Node) -> Any:
+                self.fake_mode, "allow_non_fake_inputs", True
+            ):
+                if has_tracing_context:
+                    return compiled_submod_real(*new_args, **kwargs)
```
This line is in a stack trace from the moco model in torchbench ( https://www.internalfb.com/phabricator/paste/view/P1189842447 ) . Looks like the fake tensors passed to the compiled submodule are then being passed to a deferred call of cudagraphify, which seems to expect real tensors.

eellison replied:
passes now

eellison commented:
Yes, training is handled but not inference yet. Syncing up with Brian then addressing.

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


Pass along the output strides from inductor compilation instead of those computed from FakeTensor. I think we can delete lazy_ddp with these changes.

Edit: it's non-trivial to propagate inductor strides to dynamo output strides because of things like Subclasses, aliasing deduplication, etc etc.

I have updated this pr so that we invoked aot_autograd once after compilation with fake tensors in order to propagate output strides. The actually inductor generated fn is skipped on this first call.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


Pass along the output strides from inductor compilation instead of those computed from FakeTensor. I think we can delete lazy_ddp with these changes.

Edit: it's non-trivial to propagate inductor strides to dynamo output strides because of things like Subclasses, aliasing deduplication, etc etc.

I have updated this pr so that we invoked aot_autograd once after compilation with fake tensors in order to propagate output strides. The actually inductor generated fn is skipped on this first call.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


            # Note: Returning Fake Tensors on First AOT Autograd Call
            #
            # Inductor will optimize strides of outputs when it deems it profitable.
            # For instance, converting to channels last. When we split the graph here
            # into multiple inductor compilations, we need to make sure that the
            # output strides of one compilation is appropriately passed to the subsequent
            # compilations. However, the mapping from inductor output to dynamo output
            # is non-trivial due to aot_autograd's deduping, de-aliasing, mutation, re-writing,
            # subclass handling, etc. In order to replay all this logic we set a flag such that
            # the first invocation of inductor in aot_autograd will return Fake Tensors with
            # appropriate strides. Then, all of aot autograd's runtime logic is replayed.
            # This gives us the appropriately strided outputs here which will reflect runtime strides.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


            # Note: Returning Fake Tensors on First AOT Autograd Call
            #
            # Inductor will optimize strides of outputs when it deems it profitable.
            # For instance, converting to channels last. When we split the graph here
            # into multiple inductor compilations, we need to make sure that the
            # output strides of one compilation is appropriately passed to the subsequent
            # compilations. However, the mapping from inductor output to dynamo output
            # is non-trivial due to aot_autograd's deduping, de-aliasing, mutation, re-writing,
            # subclass handling, etc. In order to replay all this logic we set a flag such that
            # the first invocation of inductor in aot_autograd will return Fake Tensors with
            # appropriate strides. Then, all of aot autograd's runtime logic is replayed.
            # This gives us the appropriately strided outputs here which will reflect runtime strides.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


            # Note: Returning Fake Tensors on First AOT Autograd Call
            #
            # Inductor will optimize strides of outputs when it deems it profitable.
            # For instance, converting to channels last. When we split the graph here
            # into multiple inductor compilations, we need to make sure that the
            # output strides of one compilation is appropriately passed to the subsequent
            # compilations. However, the mapping from inductor output to dynamo output
            # is non-trivial due to aot_autograd's deduping, de-aliasing, mutation, re-writing,
            # subclass handling, etc. In order to replay all this logic we set a flag such that
            # the first invocation of inductor in aot_autograd will return Fake Tensors with
            # appropriate strides. Then, all of aot autograd's runtime logic is replayed.
            # This gives us the appropriately strided outputs here which will reflect runtime strides.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

wconstab commented:
this sounds like a reasonable approach to me.

i'm confused about how non-lazy ddp optimizer works now.  it looks like it runs as a compile_fn given example inputs.  wouldn't this be considered lazy?

bdhirsh commented on torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py:
```diff
@@ -98,9 +114,16 @@ def aot_dispatch_base(
+        # see note: [Returning Fake Tensors on First AOT Autograd Call]
+        if tracing_context and tracing_context.fakify_first_call:
+            fakified_out = _compute_output_meta_with_inductor_strides(
+                fw_module, fwd_output_strides
```
hmm so the invariant here is that:

(1) `tracing_context.fakify_first_call` is only true (today) while ddp_optimizer is in the middle of tracing

(2) fwd_output_strides is only set (today) if inductor was the backend.

So this code only does stuff when the two are used together

eellison replied:
fakified_out is still set even if inductor is not the backend, it will just be the default meta strides.

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -68,6 +68,7 @@ def create_runtime_wrapper(
     trace_joint: bool,
     keep_input_mutations: bool,
     disable_amp: bool,
+    fakified_first_outputs: Optional[List[Tensor]] = None,
```
bleh - today, this `create_runtime_wrapper()` is meant to be (mostly) agnostic to training vs. inference. So it's pretty confusing that the "fakified_first_outputs" runtime logic gets run in the runtime wrapper for inference, but not for training.

I think I'd rather just manually shove this logic somewhere here instead https://github.com/pytorch/pytorch/pull/120523/files#diff-cb8e02fc8f37e53904ab1b151c46dd109cf50d8121bbd340834b2e976b22ebc4R159 (so the `aot_dispatch_base` and `aot_dispatch_autograd` paths both have some shared helper fn they can call). Does that sound ok?

eellison replied:
Sounds good, will update

bdhirsh commented on torch/_dynamo/backends/distributed.py:
```diff
@@ -311,8 +313,43 @@ def run_node(self, n: Node) -> Any:
+            # appropriate strides. Then, all of aot autograd's runtime logic is replayed.
+            # This gives us the appropriately strided outputs here which will reflect runtime strides.
+
+            class FakifyGuard:
```
rename: "FakeifyFirstAOTInvocationGuard"?

bdhirsh commented on torch/_dynamo/backends/distributed.py:
```diff
@@ -311,8 +313,43 @@ def run_node(self, n: Node) -> Any:
             compiled_submod_real = self.compile_submod(real_mod, new_args, kwargs)
 
+            # TODO - better way of doing this?
+            # Only aot autograd handles fakifying first call
```
yeah... good call.

Alternatively - do we have a way to detect if we are compiling for inductor? If we do, that might be easier than trying to add support for this for any backend that happens to use aot autograd (hopefully most other backends don't change output striding)

bdhirsh replied:
Oh maybe we could even branch here on `report_output_strides()`, since that's the only case that matters anyway

eellison replied:
At this point we have already passed in fakify_first_outputs to aot_autograd. If it were a non inductor backend then we'd still have to invoke it now so that the first output wouldn't be fake.

bdhirsh commented on torch/_dynamo/backends/distributed.py:
```diff
@@ -321,14 +358,20 @@ def run_node(self, n: Node) -> Any:
+                self.fake_mode, "allow_non_fake_inputs", True
+            ):
+                if has_tracing_context and invoked_aot_autograd:
+                    return compiled_submod_real(*new_args, **kwargs)
```
Do you think we could figure out a way to add some extra assertions and make this even more explicit?

One option:
(1) Manually use the fake mode to convert `*new_args, **kwargs` to be fake
(2) add a loud assert in the runtime code that asserts that whenever that `first_fake_invocation` is set at runtime, all of our tensor args are fake.

Mostly just trying to guard against the error case where a user sends us their runtime inputs, and we accidentally run this new code path and return them fake tensors.

eellison replied:
> (1) Manually use the fake mode to convert *new_args, **kwargs to be fake
 
 They're fake at this point. I thought I needed the `allow_non_fake` at one point because of a test failure where the parameters weren't being updated to fake. but i just reran tests and didnt get a failure so let me try to remove.

2) sure - can add assertion. in the runtime code do you mean in aot_autograd ?

bdhirsh replied:
> They're fake at this point.

ah cool


>  in the runtime code do you mean in aot_autograd ?

yep - in the actual aot autograd runtime code (that secretly returns the FakeTensor outputs) - just an assert that all of the "runtime inputs" are also fake would be nice.

eellison replied:
The parameters are not actually fake, and the outputs might be subclasses. I did add a test that none of the outputs are torch.Tensors though.

bdhirsh replied:
There’s also an is_fake method in fake_tensor.py that handles subclasses properly (assert all their inner tensors are fake)

bdhirsh commented:
> i'm confused about how non-lazy ddp optimizer works now. it looks like it runs as a compile_fn given example inputs. wouldn't this be considered lazy?

Elias's note here https://github.com/pytorch/pytorch/pull/120523/files#diff-a885ebf62b3ac8ad62acdbca474dff7e7a7daa26d28f88278bfd8514d3168c4cR317 explains it more, but:

ddp_optimizer is not actually "invoking the runtime (inductor-generated) code". Elias just needs a way to get the (complicated) output tensor calling convention between aot autograd and dynamo. And there isn't really a good way to do this, so instead he is just manually invoking the AOTAutograd, and letting the runtime glue code perform the calling convention changes so he can grab the (FakeTensor) dynamo outputs with proper inductor strides.

bdhirsh reviewed (approved):
left some nits! (I think the extra assertion in particular would be good to have)

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


            # Note: Returning Fake Tensors on First AOT Autograd Call
            #
            # Inductor will optimize strides of outputs when it deems it profitable.
            # For instance, converting to channels last. When we split the graph here
            # into multiple inductor compilations, we need to make sure that the
            # output strides of one compilation is appropriately passed to the subsequent
            # compilations. However, the mapping from inductor output to dynamo output
            # is non-trivial due to aot_autograd's deduping, de-aliasing, mutation, re-writing,
            # subclass handling, etc. In order to replay all this logic we set a flag such that
            # the first invocation of inductor in aot_autograd will return Fake Tensors with
            # appropriate strides. Then, all of aot autograd's runtime logic is replayed.
            # This gives us the appropriately strided outputs here which will reflect runtime strides.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


            # Note: Returning Fake Tensors on First AOT Autograd Call
            #
            # Inductor will optimize strides of outputs when it deems it profitable.
            # For instance, converting to channels last. When we split the graph here
            # into multiple inductor compilations, we need to make sure that the
            # output strides of one compilation is appropriately passed to the subsequent
            # compilations. However, the mapping from inductor output to dynamo output
            # is non-trivial due to aot_autograd's deduping, de-aliasing, mutation, re-writing,
            # subclass handling, etc. In order to replay all this logic we set a flag such that
            # the first invocation of inductor in aot_autograd will return Fake Tensors with
            # appropriate strides. Then, all of aot autograd's runtime logic is replayed.
            # This gives us the appropriately strided outputs here which will reflect runtime strides.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


            # Note: Returning Fake Tensors on First AOT Autograd Call
            #
            # Inductor will optimize strides of outputs when it deems it profitable.
            # For instance, converting to channels last. When we split the graph here
            # into multiple inductor compilations, we need to make sure that the
            # output strides of one compilation is appropriately passed to the subsequent
            # compilations. However, the mapping from inductor output to dynamo output
            # is non-trivial due to aot_autograd's deduping, de-aliasing, mutation, re-writing,
            # subclass handling, etc. In order to replay all this logic we set a flag such that
            # the first invocation of inductor in aot_autograd will return Fake Tensors with
            # appropriate strides. Then, all of aot autograd's runtime logic is replayed.
            # This gives us the appropriately strided outputs here which will reflect runtime strides.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison committed:
Update on "Pass inductor strides forward in ddp optimizer"


            # Note: Returning Fake Tensors on First AOT Autograd Call
            #
            # Inductor will optimize strides of outputs when it deems it profitable.
            # For instance, converting to channels last. When we split the graph here
            # into multiple inductor compilations, we need to make sure that the
            # output strides of one compilation is appropriately passed to the subsequent
            # compilations. However, the mapping from inductor output to dynamo output
            # is non-trivial due to aot_autograd's deduping, de-aliasing, mutation, re-writing,
            # subclass handling, etc. In order to replay all this logic we set a flag such that
            # the first invocation of inductor in aot_autograd will return Fake Tensors with
            # appropriate strides. Then, all of aot autograd's runtime logic is replayed.
            # This gives us the appropriately strided outputs here which will reflect runtime strides.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison committed:
Update

[ghstack-poisoned]

kadeng commented:
I think we should add some tests where we explicitly test this with keep_output_stride=False. At the moment, by default, keep_output_stride is True. Should not block this PR, though, we could do this in an additional followup PR. We also might want to have an experimental (draft) PR stacked on top where we set keep_output_stride=False by default and see what happens in a full CI run.

eellison commented:
@kadeng  we are testing it in `_test_graph_split_inductor_layout_optimizations_impl`

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge
